### ISSUE TITLE
Hoeffding Tree fails intermittently

### ISSUE THREAD

**Issue by dkipke** [mlpack/mlpack#2825]
#### Issue description

The HoeffdingTree function fails intermittently.  It seems to execute correctly some times, but it gets a segmentation fault other times.  I've  ad hoc tried different parameters to see if I could discern the issue, but that has not revealed anything.  With verbose=true, the logs indicate that the function returns if it successfully classifies at least one sample, but fails if it does not.    

#### Your environment

 * version of mlpack: not sure how to tell.  I built from the master branch pulled on Dec 22
 * operating system:  MacOS 10.15.7
 * compiler:  gcc compiler (I think)
 * version of dependencies (Boost/Armadillo):
 * any other environment information you think is relevant:
Use the Go bindings

#### Steps to reproduce

here are log messages:   It executes in the first instance and fails on the second instance.  

====== instance 1 =========
ml_lib_core/Hoeffding TRAIN: NtvChanIdx=0, param.Lib={BatchMode:false Bins:10 Confidence:0.9 InfoGain:true InputModel:<nil> Labels:0xc000012240 MaxSamples:5000 MinSamples:10 NumericSplitStrategy:binary ObservationsBeforeBinning:1 Passes:10 Test:<nil> TestLabels:<nil> Training:0xc000142460 Verbose:true} 
  !! idx=0: training data shape=1, 20, labels=1, 1
  !! idx=0: labels=[0] 
[INFO ] 0 mappings in dimension 0.
[INFO ] 0 mappings in dimension 1.
[INFO ] 0 mappings in dimension 2.
[INFO ] 0 mappings in dimension 3.
[INFO ] 0 mappings in dimension 4.
[INFO ] 0 mappings in dimension 5.
[INFO ] 0 mappings in dimension 6.
[INFO ] 0 mappings in dimension 7.
[INFO ] 0 mappings in dimension 8.
[INFO ] 0 mappings in dimension 9.
[INFO ] 0 mappings in dimension 10.
[INFO ] 0 mappings in dimension 11.
[INFO ] 0 mappings in dimension 12.
[INFO ] 0 mappings in dimension 13.
[INFO ] 0 mappings in dimension 14.
[INFO ] 0 mappings in dimension 15.
[INFO ] 0 mappings in dimension 16.
[INFO ] 0 mappings in dimension 17.
[INFO ] 0 mappings in dimension 18.
[INFO ] 0 mappings in dimension 19.
[INFO ] 1 out of 1 correct on training set (100).
[INFO ] 1 nodes in the tree.
  !!   ==> TRAIN: idx=0, POST CALL OK: param.Lib={false 10 0.9 true <nil> 0xc000012240 5000 10 binary 1 10 <nil> <nil> 0xc000142460 true}

====== instance 2 =========
!! ml_lib_core/Hoeffding TRAIN: NtvChanIdx=1, param.Lib={BatchMode:false Bins:10 Confidence:0.9 InfoGain:true InputModel:<nil> Labels:0xc000012380 MaxSamples:5000 MinSamples:10 NumericSplitStrategy:binary ObservationsBeforeBinning:1 Passes:10 Test:<nil> TestLabels:<nil> Training:0xc0001424a0 Verbose:true} 
  !! idx=1: training data shape=204, 20, labels=204, 1
  !! idx=1: labels=[1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 2 0 0 1 1 1 0 1 1 1 1 2 2 1 1 1 0 1 1 0 0 1 0 1 0 1 1 1 1 0 1 0 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 0 1 1 0 0 1 1 2 2 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 0 1 0 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 1 1 0 1 1 0 1 2 1 0 1 1 1 1 1 1 1 2 1 0 0 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 0 0 1 1 0 1 1 1 1 1 1 2 0 0 0 0 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 0 1 0 1 1 1 1 0 1 0 1] 
[INFO ] 0 mappings in dimension 0.
[INFO ] 0 mappings in dimension 1.
[INFO ] 0 mappings in dimension 2.
[INFO ] 0 mappings in dimension 3.
[INFO ] 0 mappings in dimension 4.
[INFO ] 0 mappings in dimension 5.
[INFO ] 0 mappings in dimension 6.
[INFO ] 0 mappings in dimension 7.
[INFO ] 0 mappings in dimension 8.
[INFO ] 0 mappings in dimension 9.
[INFO ] 0 mappings in dimension 10.
[INFO ] 0 mappings in dimension 11.
[INFO ] 0 mappings in dimension 12.
[INFO ] 0 mappings in dimension 13.
[INFO ] 0 mappings in dimension 14.
[INFO ] 0 mappings in dimension 15.
[INFO ] 0 mappings in dimension 16.
[INFO ] 0 mappings in dimension 17.
[INFO ] 0 mappings in dimension 18.
[INFO ] 0 mappings in dimension 19.
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6e11940]

**Comment by rcurtin** [comment]
Do you think you could also provide the data that you used for training so that we can reproduce the issue exactly?

**Comment by dkipke** [comment]
Here is the output of a unit test using a small dataset

// === These were back-to-back calls in a unit test using randomly generated data samples =====

// CALL 1 -- HoeffdingTree() call OK
param.Lib = {BatchMode:false Bins:10 Confidence:0.99 InfoGain:true InputModel:<nil> Labels:0xc000045480 MaxSamples:5000 MinSamples:10 NumericSplitStrategy:binary ObservationsBeforeBinning:1 Passes:10 Test:<nil> TestLabels:<nil> Training:0xc00000f280 Verbose:true} 
training data shape=[10, 5],  labels shape=[10, 1]
labels=[1 2 1 2 1 1 1 3 2 2] 
k=0: sample=[0.13848027479201822 0.21328568010005095 0.33231069111939204 0.3785013850537628 0.5202027747111361]
k=1: sample=[-0.1271043843427181 -0.21535548494617124 -0.4394526294976885 -0.2911366222671663 -0.4395143427585243]
k=2: sample=[0.05890336424808665 0.2872998696899754 0.18200781319489212 0.2659638474710884 0.4690522606418968]
k=3: sample=[-0.0476345425951938 -0.2830199847491545 -0.3383167217123779 -0.3618755756303934 -0.5223297484173438]
k=4: sample=[-0.060887425014108015 0.25358470573417574 0.38256645962516467 0.6436483254786712 0.5512210023577618]
k=5: sample=[0.05572227405739924 0.19374781395692992 0.19847002627993865 0.3792543276570749 0.41806017312657523]
k=6: sample=[0.002447765504816249 0.1777424956360386 0.3648368450361845 0.5438657410004102 0.5782900515547608]
k=7: sample=[-0.35625630237588857 0.2867249094854489 -0.4053441952732963 0.31327072511617493 -0.5933212732834622]
k=8: sample=[-0.14074119786533973 -0.11238661478163864 -0.4230017320722218 0.005930297014185015 -0.5506178443041072]
k=9: sample=[-0.05559756347675421 -0.16548090821694594 -0.5018475692913458 -0.2489322090313149 -0.5052986727753556]
[INFO ] 0 mappings in dimension 0.
[INFO ] 0 mappings in dimension 1.
[INFO ] 0 mappings in dimension 2.
[INFO ] 0 mappings in dimension 3.
[INFO ] 0 mappings in dimension 4.
[INFO ] 5 out of 10 correct on training set (50).
[INFO ] 1 nodes in the tree.
==> TRAIN: HoeffdingTree call OK: 
returned param.Lib={false 10 0.99 true <nil> 0xc000045480 5000 10 binary 1 10 <nil> <nil> 0xc00000f280 true}

// CALL 2 -- HoeffdingTree() call OK
param.Lib = {BatchMode:false Bins:10 Confidence:0.95 InfoGain:true InputModel:0xc000011f28 Labels:<nil> MaxSamples:5000 MinSamples:10 NumericSplitStrategy:binary ObservationsBeforeBinning:1 Passes:1 Test:0xc00000f400 TestLabels:<nil> Training:<nil> Verbose:true} 
test data shape=[10, 5] 
k=0: sample=[0.13848027479201822 0.21328568010005095 0.33231069111939204 0.3785013850537628 0.5202027747111361]
k=1: sample=[-0.1271043843427181 -0.21535548494617124 -0.4394526294976885 -0.2911366222671663 -0.4395143427585243]
k=2: sample=[0.05890336424808665 0.2872998696899754 0.18200781319489212 0.2659638474710884 0.4690522606418968]
k=3: sample=[-0.0476345425951938 -0.2830199847491545 -0.3383167217123779 -0.3618755756303934 -0.5223297484173438]
k=4: sample=[-0.060887425014108015 0.25358470573417574 0.38256645962516467 0.6436483254786712 0.5512210023577618]
k=5: sample=[0.05572227405739924 0.19374781395692992 0.19847002627993865 0.3792543276570749 0.41806017312657523]
k=6: sample=[0.002447765504816249 0.1777424956360386 0.3648368450361845 0.5438657410004102 0.5782900515547608]
k=7: sample=[-0.35625630237588857 0.2867249094854489 -0.4053441952732963 0.31327072511617493 -0.5933212732834622]
k=8: sample=[-0.14074119786533973 -0.11238661478163864 -0.4230017320722218 0.005930297014185015 -0.5506178443041072]
k=9: sample=[-0.05559756347675421 -0.16548090821694594 -0.5018475692913458 -0.2489322090313149 -0.5052986727753556]
[INFO ] 1 nodes in the tree.
==> VALIDATION: HoeffdingTree call OK: 
returned param.Lib={false 10 0.95 true 0xc000011f28 <nil> 5000 10 binary 1 1 0xc00000f400 <nil> <nil> true}

// CALL 3  -- HoeffdingTree() call FAILED
param.Lib = {BatchMode:false Bins:10 Confidence:0.99 InfoGain:true InputModel:0xc000011f30 Labels:<nil> MaxSamples:5000 MinSamples:10 NumericSplitStrategy:binary ObservationsBeforeBinning:1 Passes:1 Test:0xc00000f6a0 TestLabels:<nil> Training:<nil> Verbose:true} 
test data shape=[10, 5] 
k=0: sample=[-0.1542709563727684 -0.23344316632494025 -0.2386818721992066 -0.3433570566232891 -0.6207649328002018]
k=1: sample=[-0.11062737585228966 -0.21619648511338307 -0.2953951289956539 -0.4025189982683469 -0.5153990945284205]
k=2: sample=[-0.13438891676274503 0.012139637223347488 -0.38302243323868346 0.5248863263932506 -0.4997746776168737]
k=3: sample=[-0.12337178850307096 -0.08670503429673049 -0.19158213680573058 0.31710620125794764 -0.42143754664791866]
k=4: sample=[-0.19364783033084132 -0.4910925438726846 -0.3191892053721226 -0.2958950126580122 -0.5324620273509166]
k=5: sample=[0.279986073611208 0.3101611508524759 0.3031904102942747 0.40384629463166244 0.5781914590095214]
k=6: sample=[-0.12587116454468691 0.13990303485712519 -0.28335266047681457 0.3532363913861883 -0.5651213690787728]
k=7: sample=[-0.05199785468429219 0.28193242916890315 -0.17004077743668047 0.41781538063455254 -0.426265387454242]
k=8: sample=[0.07097375233542139 0.16941597851286355 0.300696854825657 0.5759471273892048 0.4964836346942347]
k=9: sample=[-0.24086799255019573 0.0744249224566807 -0.2101812658522384 0.3915753379944849 -0.6012324044161139]
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6bff8b4]

runtime stack:
runtime.throw(0x475267c, 0x2a)
	/usr/local/go/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:726 +0x48c

goroutine 6 [syscall]:
runtime.cgocall(0x45eb090, 0xc0001d7438, 0xd)
	/usr/local/go/src/runtime/cgocall.go:133 +0x5b fp=0xc0001d7408 sp=0xc0001d73d0 pc=0x400615b
mlpack.org/v1/mlpack._Cfunc_mlpackHoeffdingTree()
	_cgo_gotypes.go:1033 +0x45 fp=0xc0001d7438 sp=0xc0001d7408 pc=0x45cb965
mlpack.org/v1/mlpack.HoeffdingTree(0xc0001d7ef0, 0xc000010018, 0x47415d8, 0x10)
	/Users/dkipke/go/src/mlpack.org/v1/mlpack/hoeffding_tree.go:257 +0x1fc fp=0xc0001d74d0 sp=0xc0001d7438 pc=0x45cdd7c
radix/internal/ml_lib/ml_lib_core.HoeffdingTree(0xc0003c8b60, 0x3fefae147ae147ae, 0xc0001d7980, 0xc000011f00, 0x0)
	/Users/dkipke/Documents/dev/sapiens/core_packages/radix/go/internal/ml_lib/ml_lib_core/core.go:120 +0x48d fp=0xc0001d7728 sp=0xc0001d74d0 pc=0x45e0ead
radix/internal/ml_lib/ml_lib_core.TestHoeffdingTree(0xc0002bb080)
	/Users/dkipke/Documents/dev/sapiens/core_packages/radix/go/internal/ml_lib/ml_lib_core/core_test.go:292 +0x2046 fp=0xc0001d7f80 sp=0xc0001d7728 pc=0x45e7446
testing.tRunner(0xc0002bb080, 0x47621c0)
	/usr/local/go/src/testing/testing.go:1123 +0xef fp=0xc0001d7fd0 sp=0xc0001d7f80 pc=0x4114d0f
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc0001d7fd8 sp=0xc0001d7fd0 pc=0x40713e1
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x2b3

**Comment by rcurtin** [comment]
Hey @dkipke, do you think you could provide some example code that I can run to reproduce the issue, then I can start debugging?  Thanks! :+1:
